php/class-settings.php

```diff
@@ -1,270 +1,597 @@
+			self::META_KEYS['storage'] => $slug,
 		);
-		$this->handles[ $page_handle ] = $setting;
-		$setting->set_param( 'page_handle', $page_handle );
+		$params         = wp_parse_args( $params, $default_params );
+		$parts          = explode( $this->separator, trim( $slug, $this->separator ) );
+		$storage_paths  = explode( $this->separator, trim( $params[ self::META_KEYS['storage'] ], $this->separator ) );
+		$path           = array();
+		$value          = array();
+		$storage        = array();
+		$last_child     = null;
+
+		// If we have an option_name, in a single field, lets set the storage name for that item only.
+		if ( ! empty( $params['option_name'] ) ) {
+			array_pop( $storage_paths ); // Knockoff the end slug.
+			$storage_key      = $this->get_storage_key( $params['option_name'], $params['type'] );
+			$storage_paths[0] = $storage_key; // Set the base storage.
+		}
 
-		add_action( "load-{$page_handle}", array( $this, 'set_active_setting' ) );
-		// Setup the Child page handles.
-		foreach ( $setting->get_settings() as $sub_setting ) {
-			if ( 'page' !== $sub_setting->get_param( 'type' ) || ! apply_filters( "cloudinary_settings_enabled_{$sub_setting->get_slug()}", true ) ) {
-				continue;
+		while ( ! empty( $parts ) ) {
+			$path[] = array_shift( $parts );
+			if ( ! empty( $storage_paths ) ) {
+				$storage[] = array_shift( $storage_paths );
+			}
+			if ( empty( $parts ) ) {
+				$value = $default;
+			}
+			$name                                 = implode( $this->separator, $path );
+			$params[ self::META_KEYS['storage'] ] = implode( $this->separator, $storage );
+			$child                                = $this->register( $name, $value, $params );
+			if ( is_wp_error( $child ) ) {
+				return $child;
 			}
-			$sub_setting->set_param( 'page_header', $setting->get_param( 'page_header' ) );
-			$sub_setting->set_param( 'page_footer', $setting->get_param( 'page_footer' ) );
-			$capability                    = $sub_setting->get_param( 'capability', $setting->get_param( 'capability' ) );
-			$page_handle                   = add_submenu_page(
-				$setting->get_slug(),
-				$sub_setting->get_param( 'page_title', $setting->get_param( 'page_title' ) ),
-				$sub_setting->get_param( 'menu_title', $sub_setting->get_param( 'page_title' ) ),
-				$capability,
-				$sub_setting->get_slug(),
-				$render_function,
-				$sub_setting->get_param( 'position' )
-			);
-			$this->handles[ $page_handle ] = $sub_setting;
-			$sub_setting->set_param( 'page_handle', $page_handle );
-
-			add_action( "load-{$page_handle}", array( $this, 'set_active_setting' ) );
+
+			if ( $last_child ) {
+				$last_child->add( $child );
+			}
+			$last_child = $child;
 		}
+
+		return $this->settings[ $slug ];
 	}
 
 	/**
-	 * Render a page.
+	 * Register a new setting with internals.
+	 *
+	 * @param string $slug    The setting slug.
+	 * @param mixed  $default The default value.
+	 * @param array  $params  The params.
+	 *
+	 * @return mixed|Setting|\WP_Error
```

Comment:
@DavidCramer 

Just to confirm if can we get a WP_Error in return?

Reply:
@pereirinha - nope, it should be removed.